src/sd.jl

```diff
@@ -768,3 +768,97 @@ function build_variable(
     x = _vectorize_variables(_error, variables)
     return VariablesConstrainedOnCreation(x, set, SymmetricMatrixShape(n))
 end
+
+"""
+    TupleShape(args::Pair{<:AbstractShape,Int}...) <: AbstractShape
+
+A shape for representing tuple-valued functions, where each element of the tuple
+is a different [`AbstractShape`](@ref) mapped to the number of scalar elements
+in that shape.
+
+## Example
+
+```jldoctest
+julia> shape = TupleShape(ScalarShape() => 1, SymmetricMatrixShape(2) => 3);
+```
+"""
+struct TupleShape{T} <: AbstractShape
+    shapes::T
+    TupleShape(args::Pair{<:AbstractShape,Int}...) = new{typeof(args)}(args)
```

Comment:
Why not a tuple of shapes ?

Reply:
We need to be able to tell the dimension of each component when converting back from the vector form to the tuple form.

Reply:
Can't the dimension be inferred from the shape ? I would rather do that than store the same info twice with the risk of inconsistency

Reply:
For scalar and the matrix shapes, yes. `VectorShape` doesn't have the information.

So I guess we could drop the dimension, but we'd need to limit things to exactly one `VectorShape` in the tuple. It seemed simpler just to record the input dimensions.